packages/hir/src/types.candy

```diff
@@ -1,53 +1,119 @@
 use ..declarations
 
-public trait HirType
+public trait HirInlineType
+
+
+public class HirErrorType {}
+impl HirErrorType: HirInlineType & Equals & Hash {
+  fun equals(other: This): Bool { true }
+  fun hash<T>(hasher: Hasher<T>) {}
+}
 
 
 public class HirNamedType {
-  public let declaration: HirTrait | HirClass
+  public let declaration: HirTrait | HirType
+  public let parameterTypes: List<HirInlineType>
+}
+impl HirNamedType: HirInlineType & Equals & Hash {
+  fun equals(other: This): Bool {
+    declaration == other.declaration
+      && (parameterTypes as Iterable<HirInlineType>)
+        .unsafeEquals(other.parameterTypes as Iterable<HirInlineType>)
+  }
+  fun hash<T>(hasher: Hasher<T>) {
+    (declaration as Hash).hash<T>(hasher)
+    parameterTypes.unsafeHash<T>(hasher)
+  }
+}
+
+
+public class HirFunctionType {
+  public let receiverType: Maybe<HirInlineType>
+  public let parameterTypes: List<HirInlineType>
+  public let returnType: HirInlineType
+}
+impl HirFunctionType: HirInlineType & Equals & Hash {
+  fun equals(other: This): Bool {
+    (receiverType as Equals) == (other.receiverType as Equals)
+      && (parameterTypes as Iterable<HirInlineType>)
+        .unsafeEquals(other.parameterTypes as Iterable<HirInlineType>)
+      && (returnType as Equals) == (other.returnType as Equals)
+  }
+  fun hash<T>(hasher: Hasher<T>) {
+    (receiverType as Hash).hash<T>(hasher)
+    parameterTypes.unsafeHash<T>(hasher)
+    (returnType as Hash).hash<T>(hasher)
+  }
 }
-impl HirNamedType: HirType
 
 
 public class HirTupleType {
-  public let types: List<HirType>
+  public let types: List<HirInlineType>
+}
+impl HirTupleType: HirInlineType & Equals & Hash {
+  fun equals(other: This): Bool {
+    (types as Iterable<HirInlineType>).unsafeEquals(other.types as Iterable<HirInlineType>)
+  }
+  fun hash<T>(hasher: Hasher<T>) { types.unsafeHash<T>(hasher) }
 }
-impl HirTupleType: HirType
 
 
-public class HirFunctionType {
-  public let receiverType: Maybe<HirType>
-  public let parameterTypes: List<HirType>
-  public let returnType: HirType
+public class HirNamedTupleType {
+  public let types: Map<String, HirInlineType>
+}
+impl HirNamedTupleType: HirInlineType & Equals & Hash {
+  fun equals(other: This): Bool { types.unsafeEquals(other.types) }
+  fun hash<T>(hasher: Hasher<T>) { types.unsafeHash<T>(hasher) }
 }
-impl HirFunctionType: HirType
 
 
-public class HirUnionType {
-  public let types: List<HirType>
-  // TODO(JonasWanke): Use a `Set<HirType>` when we have sets that keep insertion order.
+public class HirEnumType {
+  public let variants: Map<String, HirInlineType>
```

Comment:
Maybe use a list to keep the order, e.g., when displaying this type in a tooltip?